core/variant/typed_dictionary.h

```diff
@@ -0,0 +1,342 @@
+		ERR_FAIL_COND_MSG(!is_same_typed(p_dictionary), "Cannot assign a dictionary with a different element type.");
+		Dictionary::operator=(p_dictionary);
+	}
+	_FORCE_INLINE_ TypedDictionary(const Variant &p_variant) :
+			TypedDictionary(Dictionary(p_variant)) {
+	}
+	_FORCE_INLINE_ TypedDictionary(const Dictionary &p_dictionary) {
+		set_typed(Variant::OBJECT, K::get_class_static(), Variant(), Variant::OBJECT, V::get_class_static(), Variant());
+		if (is_same_typed(p_dictionary)) {
+			Dictionary::operator=(p_dictionary);
+		} else {
+			assign(p_dictionary);
+		}
+	}
+	_FORCE_INLINE_ TypedDictionary() {
+		set_typed(Variant::OBJECT, K::get_class_static(), Variant(), Variant::OBJECT, V::get_class_static(), Variant());
+	}
+};
+
+template <typename K, typename V>
+struct VariantInternalAccessor<TypedDictionary<K, V>> {
+	static _FORCE_INLINE_ TypedDictionary<K, V> get(const Variant *v) { return *VariantInternal::get_dictionary(v); }
+	static _FORCE_INLINE_ void set(Variant *v, const TypedDictionary<K, V> &p_dictionary) { *VariantInternal::get_dictionary(v) = p_dictionary; }
+};
+
+template <typename K, typename V>
+struct VariantInternalAccessor<const TypedDictionary<K, V> &> {
+	static _FORCE_INLINE_ TypedDictionary<K, V> get(const Variant *v) { return *VariantInternal::get_dictionary(v); }
+	static _FORCE_INLINE_ void set(Variant *v, const TypedDictionary<K, V> &p_dictionary) { *VariantInternal::get_dictionary(v) = p_dictionary; }
+};
+
+template <typename K, typename V>
+struct PtrToArg<TypedDictionary<K, V>> {
+	_FORCE_INLINE_ static TypedDictionary<K, V> convert(const void *p_ptr) {
+		return TypedDictionary<K, V>(*reinterpret_cast<const Dictionary *>(p_ptr));
+	}
+	typedef Dictionary EncodeT;
+	_FORCE_INLINE_ static void encode(TypedDictionary<K, V> p_val, void *p_ptr) {
+		*(Dictionary *)p_ptr = p_val;
+	}
+};
+
+template <typename K, typename V>
+struct PtrToArg<const TypedDictionary<K, V> &> {
+	typedef Dictionary EncodeT;
+	_FORCE_INLINE_ static TypedDictionary<K, V>
+	convert(const void *p_ptr) {
+		return TypedDictionary<K, V>(*reinterpret_cast<const Dictionary *>(p_ptr));
+	}
+};
+
+template <typename K, typename V>
+struct GetTypeInfo<TypedDictionary<K, V>> {
+	static const Variant::Type VARIANT_TYPE = Variant::DICTIONARY;
+	static const GodotTypeInfo::Metadata METADATA = GodotTypeInfo::METADATA_NONE;
+	static inline PropertyInfo get_class_info() {
+		return PropertyInfo(Variant::DICTIONARY, String(), PROPERTY_HINT_DICTIONARY_TYPE, vformat("%s;%s", K::get_class_static(), V::get_class_static()));
+	}
+};
+
+template <typename K, typename V>
+struct GetTypeInfo<const TypedDictionary<K, V> &> {
+	static const Variant::Type VARIANT_TYPE = Variant::DICTIONARY;
+	static const GodotTypeInfo::Metadata METADATA = GodotTypeInfo::METADATA_NONE;
+	static inline PropertyInfo get_class_info() {
+		return PropertyInfo(Variant::DICTIONARY, String(), PROPERTY_HINT_DICTIONARY_TYPE, vformat("%s;%s", K::get_class_static(), V::get_class_static()));
```

Comment:
```suggestion
		return PropertyInfo(Variant::DICTIONARY, String(), PROPERTY_HINT_DICTIONARY_TYPE, vformat("%s,%s", K::get_class_static(), V::get_class_static()));
```

Are there any considerations why this should be `;` and not `,`? It's less confusing than having both a comma and a semicolon as a separator in different places.

Reply:
Enums already use commas in their hint strings; opting for a semicolon prevents overlap between the two. It's fine to make the separator a comma once the enum data is parsed and discarded, when the representation is no longer a hint string

Reply:
There is a difference between `PROPERTY_HINT_ARRAY_TYPE` and `PROPERTY_HINT_TYPE_STRING`. This causes some confusion.

* The former uses the type name as a string and does not support specifying hint and hint string elements. Hint string examples:
  * `"int"` - `Array[int]`;
  * `"Node2D"` - `Array[Node2D]`;
  * `"Image"` - `Array[Image]`.
* The latter encodes the Variant type of an element, its hint and hint string using `:` and `/`. Hint string examples:
  * `"%d:" % [TYPE_INT]` - `Array[int]`;
  * `"%d/%d:%s" % [TYPE_OBJECT, PROPERTY_HINT_NODE_TYPE, "Node2D"]` - `Array[Node2D]`;
  * `"%d/%d:%s" % [TYPE_OBJECT, PROPERTY_HINT_RESOURCE_TYPE, "Image"]` - `Array[Image]`.

`PROPERTY_HINT_TYPE_STRING` [allows](https://docs.godotengine.org/en/stable/classes/class_%40globalscope.html#class-globalscope-constant-property-hint-type-string) you to encode nested array types, even though they are not yet supported in GDScript.

Accordingly, for dictionaries we should also have different contents for `PROPERTY_HINT_DICTIONARY_TYPE` and `PROPERTY_HINT_TYPE_STRING`. For example, to encode `Dictionary[TKey, TValue]` you can use:

* `PROPERTY_HINT_DICTIONARY_TYPE`:
  * `TKeyString,TValueString`
* `PROPERTY_HINT_TYPE_STRING`:
  * `{TKeyBuiltin:|TValueBuiltin:}`
  * `{TKeyBuiltin/TKeyHint:TKeyHintString|TValueBuiltin/TValueHint:TValueHintString}`

We also need to think about cases of an array nested in a dictionary, a dictionary nested in an array, etc.

Reply:
It's insane how much of a mess this whole system is but let's tap our ears for this minor release.

Reply:
I haven't messed with `PROPERTY_HINT_TYPE_STRING` at all, so that's something I'll have to dig into. Honestly, I wasn't even aware that it could be used for setting up typed arrays; I always thought it was for enum-specific contexts.

And yes, this entire system is in need of a rework. However, the objective of this PR has always been to setup typed dictionaries with the existing system, warts and all.

Reply:
Regarding `PROPERTY_HINT_TYPE_STRING`, I think it's not urgent, we can figure it out later. The only thing that should be decided in this PR is what separator to use: comma or semicolon.

Reply:
When actually displayed to the user, the separator will be `, `. If we're talking internally, I suppose we could make the separator something universal. I'm not married to any one format, as I only went with `;` for the sake of enum compatibility.

Reply:
Gave this some thought/tests, and this might just have to *stay* messy to some extent, at least if we're wanting to keep compatibility with enum string hints. For now though, the internal separator is `;`, with `, ` being used for the GDScript type hint & user-visible contexts. Not gonna fuss too much over the semantics, as a typesystem rework would break compatibility anyway.